[mlir] Add metadata to Diagnostic.

weiweichen opened the pull request:
Add metadata to Diagnostic. 

Motivation: we have a use case where we want to do some filtering in our customized Diagnostic Handler based on some customized info that is not `location` or `severity` or `diagnostic arguments` that are member variables of `Diagnostic`. Specifically, we want to add a unique ID to the `Diagnostic` for the handler to filter in a compiler pass that emits errors in async tasks with multithreading and the diagnostic handling is associated to the task.

This patch adds a field of `metadata` to `mlir::Diagnostics` as a general solution. `metadata` is of type `SmallVector<DiagnosticArgument, 0>` to save memory size and reuse existing `DiagnosticArgument` for metadata type.

weiweichen committed:
Add metadat to Diagnostic.

weiweichen committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into weiweic/add-diagnostic-metadata

joker-eph commented on mlir/include/mlir/IR/Diagnostics.h:
```diff
@@ -271,6 +271,9 @@ class Diagnostic {
   }
 
+  /// Returns the current list of diagnostic metadata.
+  SmallVector<DiagnosticArgument, 1>& getMetaData() { return metadata; }
```
```suggestion
  SmallVectorImpl<DiagnosticArgument>& getMetaData() { return metadata; }
```

joker-eph commented on mlir/include/mlir/IR/Diagnostics.h:
```diff
@@ -290,6 +293,9 @@ class Diagnostic {
   NoteVector notes;
+
+  /// A list of metadata attached to this Diagnostic.
+  SmallVector<DiagnosticArgument, 1> metadata;
```
```suggestion
  SmallVector<DiagnosticArgument, 0> metadata;
```

joker-eph reviewed (changes requested):
Thanks for the contributions:

- Can you please provide some tests? These also serve as examples for the feature.
- Can you also elaborate on the use-cases in the PR description?

weiweichen committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into weiweic/add-diagnostic-metadata

weiweichen committed:
Apply review comment and format.

weiweichen committed:
Add test for Diagnostic metadata.

weiweichen committed:
Actually add the tests.

weiweichen committed:
Update comment.

weiweichen commented:
> Thanks for the contributions:
> 

> * Can you please provide some tests? These also serve as examples for the feature.

Yep! Added `mlir/test/IR/diagnostic-handler-metadata.mlir`

> * Can you also elaborate on the use-cases in the PR description?

Yes, updated the PR description with a bit details on the downstream use case we have that needs this upstream change.

Thank you very much for the feedback!

weiweichen commented:
@joker-eph could you please take a look to see if there should be more changes with this PR? Thank you!

joker-eph commented on mlir/test/IR/diagnostic-handler-metadata.mlir:
```diff
@@ -0,0 +1,19 @@
+
+// CHECK-LABEL: Test 'test'
+// CHECK-NEXT: 8:3: error: test diagnostic metadata
+// CHECK-NOT: 13:3: error: test diagnostic metadata
```
Instead of using FileCheck, can't we use the `--verify-diagnostic` mode of mlir-opt?

weiweichen replied:
Ah, yes, I was following [what `diagnostic-handler-filter.mlir` is done for checking the diagnostics](https://github.com/llvm/llvm-project/blob/2e57e6366677390110f5382894c8afeba8da7419/mlir/test/IR/diagnostic-handler-filter.mlir#L5-L6). 

`-verify-diagnostic` is a much better way, updated the test.

joker-eph reviewed (commented):
Thanks for adding the test!

Is there some documentation to update as well?

weiweichen committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into weiweic/add-diagnostic-metadata

weiweichen committed:
Update test to use -verify-diagnostic.

weiweichen committed:
Add some documentation about diagnostic metadata.

weiweichen committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into weiweic/add-diagnostic-metadata

weiweichen commented:
> Thanks for adding the test!
> 
> Is there some documentation to update as well?

Yep, added something to `Diagnostic.md` about `metadata`.
